docs/architecture/design-proposals/helm3.md

```diff
@@ -0,0 +1,38 @@
+﻿# Kubeapps Helm 3 support
+
+We believe that the transition to Helm 3 can be done in such a fashion that
+both the old tiller-proxy and the new Helm 3 components can coexist.
+However, we propose that the initial effort not touch tiller-proxy, but create
+a completely new component.
+The choice between Helm 2 support and Helm 3 support should be made at deploy
+time in Helm (e.g. by setting `helm3=true` in Values).
+
+Since Helm 3 has gotten rid of Tiller, it instead provides a client library
+that can be used to access all required commands.
+We see no reason to implement a proxy for Helm 3, but rather what we call an
+agent.
+We should update the API version from v1 to v2, to clearly distinguish between
+the old and the new subsystems.
+That will hopefully help with compatibility later on in case the architecture
+changes more.
+
+**Current situation:**
+![Current situation](https://user-images.githubusercontent.com/7773090/67413010-ac044e00-f5c0-11e9-93e9-f3cdd1eeaca8.PNG)
+
+**With the new additions:**
+![With the new additions](https://user-images.githubusercontent.com/7773090/67413025-b45c8900-f5c0-11e9-8961-67377bc8faad.PNG)
+
+## Authentication
+
+Since Helm 2 built on Tiller, all authentication to the Kubernetes cluster
+happened "over there" and Kubeapps did not need to authenticate anything other
+than the communication with Tiller.
```

Comment:
This was not clear to me at first, but after digging a bit, I think I understand and agree, though would expand and clarify a little as follows:

> Since Helm 2 is built on the assumption that the Tiller service runs in-cluster with its own service account (as per Helm's [Securing your Helm Installation](https://helm.sh/docs/securing_installation/#tiller-and-user-permissions) document, "Tiller in its current form does not provide a way to map user credentials to specific permissions within Kubernetes. When Tiller is running inside of the cluster, it operates with the permissions of its service account."), therefore the only credentials Kubeapps was required to provide for authentication with Tiller was the `ca.crt` - and then only if tiller had been configured with tls. For this reason Kubeapps currently authenticates with the Kubernetes API server using the users bearer token not only when talking directly to the api server, but also to verify permissions, before making any request to tiller.